Map pin colors correspond with rows in results

twcarson opened the pull request:
When geomapping a cross-search, the color of the map pins now correspond to rows in the search results.  This is done by creating a mapping between lings and the rows that they appear in, and another mapping between colors and the row that they correspond to.

This fix partially remedies the existing problem that the map will stop laying pins when it runs out of colors, since there are typically many lings per row, and often fewer than 20 (max number of colors) rows.  The solution is not scalable, however, and large cross-searches still fail to display all map-pins.

twcarson committed:
Map pin colors correspond with rows in results 

When geomapping a cross-search, the color of the map pins now correspond to rows in the search results.  This is done by creating a mapping between lings and the rows that they appear in, and another mapping between colors and the row that they correspond to.

This fix partially remedies the existing problem that the map will stop laying pins when it runs out of colors, since there are typically many lings per row, and often fewer than 20 (max number of colors) rows.  The solution is not scalable, however, and large cross-searches still fail to display all map-pins.

dej611 commented on app/assets/javascripts/views/searches/results/results_cross.js:
```diff
@@ -206,16 +221,32 @@
+
+      //modified section
+      function styler(entry) {
+        if(colors_by_row[rows_by_ling[entry.id]]) {
```
Can we refactor a little bit here?

Also note that this code will work for (or up to) the top 20 result rows. And it does not provide any way to change them yet. (Perhaps add a comment about that).

Something like:

```js
// Assing a color up to 20 rows
function getColorById(id){
  if(!colors_by_row[rows_by_ling[id]]){
    if(counter < 20){
      colors_by_row[rows_by_ling[id]] = counter;
    }
  }
  return  colors_by_row[rows_by_ling[id]];
}

// Style a map pin/marker. It has a limit of 20 colors
function styler(entry) {
  counter++;
  var color = getColorById(entry.id);
  if(color == null){
    return null;
  }
  return {
    markerColor: color,
    iconColor: 'white',
    icon: 'info',
    text: popups[entry.id]
  };
}
```

dej611 replied:
Something like:

twcarson replied:
This looks good to me.  I've added this to a second commit.

dej611 commented on app/assets/javascripts/views/searches/results/results_cross.js:
```diff
@@ -26,6 +26,21 @@
+    //lings based on their row in getStyler().
+    var rows_by_ling = {}
+    if (resultsJson) {
+      var json_rows = resultsJson.rows;
```
You can use either a `.reduce` function here or a `.forEach` if you find it easier to code:

```js
resultsJson.rows.forEach( function(row, i){
  row.child.forEach( function(child){
    rows_by_ling[child.lings_property.ling_id] = i;
  });
});
```

twcarson replied:
Agree with this too.  added to changes.

twcarson committed:
"Map Pins now Colored by Row" Refactored

No changes in functionality between last commit and now.  just refactored:

 - getStyler() now written in more succinct, idiomatic JS
    + associating rows with colors is now the job of the function getColorByRow
 - Added comments detailing limitations of getStyler()
    + still stops assigning colors after 20 rows
    + this is still due to the way that map-pin colors are handled.

twcarson committed:
results_cross.js: a bit more stylistic refactoring

This could have been part of the last commit.  
Once again, no changes in functionality, 
just cleaning up style and readability.
